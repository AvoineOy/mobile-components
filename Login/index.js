import React from 'react';
import {
  Text, processColor, View, TextInput, Button, Alert
} from 'react-native';
import AvoineSSOClient from '@avoine/sso-client';

export default class AvoineSSOLogin extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      phoneNumber: '',
      isLoggingIn: false,
      showUseCode: false,
    };

    //this.render = this.render.bind(this);
    this.requestCode = this.requestCode.bind(this);
    this.useCode = this.useCode.bind(this);
  }

  requestCode() {
    this.setState({ isLoggingIn: true });
    console.log('Request code for', this.state.phoneNumber);
    this.props.ssoClient
      .requestCode(this.state.phoneNumber)
      .then(hash => {
        this.setState({ isLoggingIn: false, showUseCode: true, hash });
      })
      .catch(err => {
        this.setState({ isLoggingIn: false, showUseCode: false });
        console.error(err);
      })
    ;
  }

  useCode() {
    this.setState({ isLoggingIn: true });

    this.props.ssoClient.validateCode(this.state.code, this.state.hash)
      .then(accessToken => {
        this.setState({ isLoggingIn: false, code: undefined });
        this.props.onLogin(accessToken);
      })
      .catch(err => {
        Alert.alert('Virhe', 'Koodia ei voitu vahvistaa. Yritä uudelleen.\n\n'
        + 'Huom!\nKoodi on voimassa\nvain 5 minuuttia.');
        this.setState({ isLoggingIn: false });
      })
    ;
  }

  render() {
    if (this.state.isLoggingIn) {
      return (
        <Text style={[this.props.style.Text, this.props.style.AvoineSSOLogin.Text]}>Odota hetki, tunnistautuminen käynnissä...</Text>
      );
    } else if (this.state.showUseCode) {
      return (
        <View>
          <Text style={[this.props.style.Text, this.props.style.AvoineSSOLogin.Text]}>Syötä koodi</Text>
          <TextInput
            defaultValue={this.state.code}
            style={[this.props.style.TextInput, this.props.style.AvoineSSOLogin.TextInput]}
            onChangeText={code => this.setState({code})}
            keyboardType="numeric"
            returnKeyType="go"
            returnKeyLabel="Kirjaudu"
          />

        <Button
          color={this.props.style.AvoineSSOLogin.Button.color || this.props.style.Button.color}
          title="Kirjaudu sisään"
          onPress={this.useCode}
        />
        </View>
      );
    }

    return (
      <View>
        <Text style={[this.props.style.Text, this.props.style.AvoineSSOLogin.Text]}>Syötä puhelinnumerosi</Text>
        <TextInput
          style={[this.props.style.TextInput, this.props.style.AvoineSSOLogin.TextInput]}
          onChangeText={phoneNumber => this.setState({phoneNumber})}
          defaultValue={this.state.phoneNumber}
          keyboardType="phone-pad"
          returnKeyType="go"
          returnKeyLabel="Jatka"
        />
        <Button
          color={this.props.style.AvoineSSOLogin.Button.color || this.props.style.Button.color}
          title="Pyydä koodi"
          onPress={this.requestCode}
        />
      </View>
    );
  }
}