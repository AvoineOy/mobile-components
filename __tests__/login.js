import 'react-native'
import React from 'react'
import Login from '../Login'
import defaultStyles from '../styles'

import renderer from 'react-test-renderer';

const appConfig = {
  Login: {
    url: 'https://example.com',
    instance: 'my_instance'
  }
}

/**
 * Test rendering against ./__snapshots__/login.js.snap
 */
test('renders correctly', () => {
  const tree = renderer.create(
    <Login
      config={appConfig}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});