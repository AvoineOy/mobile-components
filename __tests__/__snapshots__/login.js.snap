// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#222",
          "fontSize": 18,
          "fontWeight": "normal",
          "lineHeight": 40,
          "textAlign": "left",
          "textDecorationLine": "none",
        },
        undefined,
      ]
    }
  >
    Syötä puhelinnumerosi
  </Text>
  <TextInput
    allowFontScaling={true}
    defaultValue=""
    keyboardType="phone-pad"
    onChangeText={[Function]}
    returnKeyLabel="Jatka"
    returnKeyType="go"
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "fontSize": 18,
          "height": 40,
          "textAlign": "center",
          "width": 300,
        },
        Object {
          "width": 300,
        },
      ]
    }
  />
  <View
    accessibilityComponentType="button"
    accessibilityLabel={undefined}
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    collapsable={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={undefined}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
            Object {
              "color": "#036cc7",
            },
          ]
        }
      >
        Pyydä koodi
      </Text>
    </View>
  </View>
</View>
`;
